# Capstone Project: Optimising Sales Strategies in a Warehouse and Retail Operations Using Data-Driven Insights.

- Introduction

This capstone project aims to analyse the "Warehouse and Retail Sales" dataset available on Data.gov to uncover trends, patterns, and opportunities for enhancing sales performance in warehouse and retail operations. The significance of this study lies in addressing a real-world business challenge: optimising sales strategies to improve profitability and operational efficiency. As e-commerce continues to grow and consumer behaviour evolves, retail and warehouse operations must adapt to remain competitive. This project seeks to provide actionable insights and data-driven solutions to support businesses in maximising their sales potential.

The project is both relevant and engaging as it integrates data analysis, business strategy, and real-world application. By examining key factors influencing sales—such as inventory management, pricing strategies, and seasonal trends—this study will offer valuable insights for businesses. The findings can contribute to informed decision-making, ultimately leading to increased revenue and improved customer satisfaction.


# ------------------------------------ DATASET ANALYSIS ---------------------------------------------

In [1]:
# import libraries:

import warnings
warnings.filterwarnings('ignore')

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns


In [2]:
# Load the dataset:

df = pd.read_csv('warehouse_and_retail_sales.csv')

In [3]:
# Display the first 10 rows:

df.head(10)

YEAR  MONTH                  SUPPLIER ITEM CODE       ITEM DESCRIPTION  \
0  2017     10        ANHEUSER BUSCH INC       175  EMPTY 1/2 KEG (30.00)   
1  2017      6        ANHEUSER BUSCH INC       175  EMPTY 1/2 KEG (30.00)   
2  2017      8        ANHEUSER BUSCH INC       175  EMPTY 1/2 KEG (30.00)   
3  2019     10  PREMIUM DISTRIBUTORS INC       175  EMPTY 1/2 KEG (30.00)   
4  2017      7        ANHEUSER BUSCH INC       175  EMPTY 1/2 KEG (30.00)   
5  2019      7  PREMIUM DISTRIBUTORS INC       175  EMPTY 1/2 KEG (30.00)   
6  2017     11        ANHEUSER BUSCH INC       175  EMPTY 1/2 KEG (30.00)   
7  2018      1        ANHEUSER BUSCH INC       175  EMPTY 1/2 KEG (30.00)   
8  2019      4  PREMIUM DISTRIBUTORS INC       175  EMPTY 1/2 KEG (30.00)   
9  2017      9        ANHEUSER BUSCH INC       175  EMPTY 1/2 KEG (30.00)   

  ITEM TYPE  RETAIL SALES  RETAIL TRANSFERS  WAREHOUSE SALES  
0   DUNNAGE           0.0               0.0          -4996.0  
1   DUNNAGE           0.0               0.0          -4804.0  
2   DUNNAGE           0.0               0.0          -4536.0  
3   DUNNAGE           0.0               0.0          -4365.0  
4   DUNNAGE           0.0               0.0          -4364.0  
5   DUNNAGE           0.0               0.0          -4349.0  
6   DUNNAGE           0.0               0.0          -4312.0  
7   DUNNAGE           0.0               0.0          -4302.0  
8   DUNNAGE           0.0               0.0          -4299.0  
9   DUNNAGE           0.0               0.0          -4210.0

In [4]:
# Display the last 10 rows:

df.tail(10)

YEAR  MONTH       SUPPLIER ITEM CODE              ITEM DESCRIPTION  \
307634  2017      7  CROWN IMPORTS     96750   CORONA EXTRA 2/12 NR - 12OZ   
307635  2019      9  CROWN IMPORTS     23445  CORONA EXTRA LOOSE NR - 12OZ   
307636  2019      6  CROWN IMPORTS     23445  CORONA EXTRA LOOSE NR - 12OZ   
307637  2019      8  CROWN IMPORTS     96750   CORONA EXTRA 2/12 NR - 12OZ   
307638  2017      8  CROWN IMPORTS     23445  CORONA EXTRA LOOSE NR - 12OZ   
307639  2017      6  CROWN IMPORTS     96750   CORONA EXTRA 2/12 NR - 12OZ   
307640  2017     11  CROWN IMPORTS     23445  CORONA EXTRA LOOSE NR - 12OZ   
307641  2019      5  CROWN IMPORTS     23445  CORONA EXTRA LOOSE NR - 12OZ   
307642  2019      7  CROWN IMPORTS     23445  CORONA EXTRA LOOSE NR - 12OZ   
307643  2020      7  CROWN IMPORTS     23445  CORONA EXTRA LOOSE NR - 12OZ   

       ITEM TYPE  RETAIL SALES  RETAIL TRANSFERS  WAREHOUSE SALES  
307634      BEER          75.0              76.0         14411.50  
307635      BEER         972.0             965.0         14511.00  
307636      BEER        1494.0            1462.0         14739.92  
307637      BEER         764.3             771.0         15289.00  
307638      BEER         990.0            1077.0         15346.71  
307639      BEER          53.0              61.0         15955.46  
307640      BEER         992.0             981.0         16079.12  
307641      BEER        1373.0            1207.0         17703.62  
307642      BEER        1198.0            1416.0         17810.87  
307643      BEER        1174.0            1041.0         18317.00

The dataset contains warehouse and retail sales records from multiple years, detailing suppliers, item descriptions and sales figures. The initial entries show "Empty 1/2 Keg" under the item type "Dunnage," with negative warehouse sales values, likely representing keg returns or inventory adjustments.The dataset appears to capture both product sales and logistical activities, such as returns and stock transfers. Given this mix of transactional data, thorough data cleaning and careful interpretation are essential to differentiate between various entry types when analysing sales trends and performance.

In [5]:
# Check basic information about the dataset:

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 307644 entries, 0 to 307643
Data columns (total 9 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   YEAR              307644 non-null  int64  
 1   MONTH             307644 non-null  int64  
 2   SUPPLIER          307478 non-null  object 
 3   ITEM CODE         307644 non-null  object 
 4   ITEM DESCRIPTION  307644 non-null  object 
 5   ITEM TYPE         307643 non-null  object 
 6   RETAIL SALES      307641 non-null  float64
 7   RETAIL TRANSFERS  307644 non-null  float64
 8   WAREHOUSE SALES   307644 non-null  float64
dtypes: float64(3), int64(2), object(4)
memory usage: 21.1+ MB


The dataset consists of 307644 entries and 9 columns related to warehouse and retail sales transactions, including key details such as year, month, supplier, item code, item description, item type, retail sales, retail transfers and warehouse sales. Most columns are completed, although some have a few missing values, particularly in the supplier and item type categories.

# Data Cleaning and Pre-processing

-Before continuing the analysis, it is essential to clean and properly format the dataset:

In [6]:
# Check for missing values:

df.isnull().sum()

YEAR                  0
MONTH                 0
SUPPLIER            166
ITEM CODE             0
ITEM DESCRIPTION      0
ITEM TYPE             1
RETAIL SALES          3
RETAIL TRANSFERS      0
WAREHOUSE SALES       0
dtype: int64

The df.isnull().sum() output shows that there are 166 missing values in the "SUPPLIER" column, 1 in "ITEM TYPE," and 3 in "RETAIL SALES," while all other columns are complete.

In [7]:
# Check for duplicates:

df.duplicated().sum()

0

the output of df.duplicated().sum() shows that there are no duplicate entries in the dataset, indicating data uniqueness across all records.

                                   # Handling missing values:

- Missing values in the 'SUPPLIER' column will be filled with 'Unknown' to maintain the total number of entries while ensuring valuable transaction data is not lost.

In [8]:
df['SUPPLIER'].fillna("UNKNOWN", inplace=True)

- The missing values in 'RETAIL SALES' will be filled with the median to prevent any skewing of the data.

In [9]:
retail_sales_median = df['RETAIL SALES'].median()
df['RETAIL SALES'].fillna(retail_sales_median, inplace=True)

- The missing values in the "ITEM TYPE" column will be removed, as the number of missing entries is minimal compared to the overall size of the dataset and deleting them will not significantly impact the analysis:

In [10]:
df.dropna(subset=['ITEM TYPE'], inplace=True)

                                   # Validate Data Types:
                                   
To ensure that the data types of each column are appropriate for analysis.

- date should be in datetime format (add new column).

- YEAR and MONTH should be integers.

- RETAIL SALES, RETAIL TRANSFERS, and WAREHOUSE SALES should be numeric (float or int).

- SUPPLIER, ITEM DESCRIPTION and ITEM TYPE should be categorical (object/string).

- ITEM CODE should be in string format, as an examination of the dataset's head revealed that some entries were displayed as numbers, all entries need to be consistent and presented in a single format rather than in two different formats.

In [11]:
# Convert date columns to datetime format and add a new date column:

df['DATE'] = pd.to_datetime(df[['YEAR', 'MONTH']].assign(DAY=1))

In [12]:
# Rename coloumns's Head for easier processing.

df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_')

In [13]:
# Standardise categorical columns for easier processing:

df['supplier'] = df['supplier'].str.strip().str.lower()
df['item_description'] = df['item_description'].str.strip().str.lower()
df['item_type'] = df['item_type'].str.strip().str.lower()

In [14]:
# Check dataset is now cleaned, pre-processed and ready for the next step of the analysis:

df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 307643 entries, 0 to 307643
Data columns (total 10 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   year              307643 non-null  int64         
 1   month             307643 non-null  int64         
 2   supplier          307643 non-null  object        
 3   item_code         307643 non-null  object        
 4   item_description  307643 non-null  object        
 5   item_type         307643 non-null  object        
 6   retail_sales      307643 non-null  float64       
 7   retail_transfers  307643 non-null  float64       
 8   warehouse_sales   307643 non-null  float64       
 9   date              307643 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(3), int64(2), object(4)
memory usage: 25.8+ MB


In [15]:
unique_values = df['item_code'].unique()
print(unique_values)

['175' 'BC' '205' ... '23886' '96750' '23445']
